File: CosineSimilarity/cosine_similarity.py
import math

def cosine_similarity(x, y):
    dot_product = sum(a*b for a,b in zip(x,y))
    norm_x = math.sqrt(sum(a*a for a in x))
    norm_y = math.sqrt(sum(b*b for b in y))
    
    if norm_x == 0 or norm_y == 0:
        return 0.0
        
    return dot_product / (norm_x * norm_y)

x = [1, 1]
y = [1, 0]

sim = cosine_similarity(x, y)
print(f"Vector x: {x}")
print(f"Vector y: {y}")
print(f"Cosine Similarity: {sim:.4f}")

File: CosineSimilarity/cosine_similarity.c
#include <stdio.h>
#include <math.h>

int main() {
    float x[] = {1.0, 1.0};
    float y[] = {1.0, 0.0};
    int n = 2;
    
    float dot = 0.0, mag_x = 0.0, mag_y = 0.0;
    
    for(int i=0; i<n; i++) {
        dot += x[i] * y[i];
        mag_x += x[i] * x[i];
        mag_y += y[i] * y[i];
    }
    
    mag_x = sqrt(mag_x);
    mag_y = sqrt(mag_y);
    
    float sim = dot / (mag_x * mag_y);
    
    printf("Vector x: (1, 1)");
    printf("Vector y: (1, 0)");
    printf("Cosine Similarity: %.4f", sim);
    
    return 0;
}